### PR TITLE
Docker images optimization / CI/CD refactoring

### PR DESCRIPTION
**Pull request by quadeare**
**Docker images optimization :**

| Image                             | Uncompressed size | Image efficiency score    | Layers    | Optimized               |
| ------                            | ------            | ------                    | ------    | ------                  |
| jolibrain/deepdetect_cpu          | 7.4 GB (New 919MB)| 57% (new 95%)             | 27 (new 18) |:+1: |
| jolibrain/deepdetect_gpu          | 12 GB (New 2.15GB)| 65% (new 95%)             | 33 (new 24) |:+1: |

CPU and GPU images contains build dependencies which increase Docker images size.
To prevent this, we use multi-stage build feature to remove all build dependencies.

**CI/CD refactoring :**

Dockerfiles duplication have been deleted and we add build argument to select cmake profile.

Build documentation have been updated on README.md file in docker folder.

### CODE REVIEW

**Comment by quadeare** [comment]
**(Proposal) Docker images name refactoring :**

Currently we have Docker images for each build type.

Example : 
* jolibrain/deepdetect_cpu
* jolibrain/deepdetect_cpu_tf
* jolibrain/deepdetect_gpu
* jolibrain/deepdetect_gpu_tf
* ...

Maybe another approach might be better using tags to specify build types.

Example (CPU) : 
* jolibrain/deepdetect:latest
* jolibrain/deepdetect:latest-caffe-tf
* jolibrain/deepdetect:v0.9.4
* jolibrain/deepdetect:v0.9.4-caffe-tf

Example (GPU) : 
* jolibrain/deepdetect:latest-gpu
* jolibrain/deepdetect:latest-gpu-caffe-tf
* jolibrain/deepdetect:v0.9.4-gpu
* jolibrain/deepdetect:v0.9.4-gpu-caffe-tf

You can checkout an example with Tensorflow project : https://hub.docker.com/r/tensorflow/tensorflow/tags

**Comment by beniz** [comment]
Hi, great work thanks! 

> * jolibrain/deepdetect:latest-gpu

One drawback I see (if I understand correctly...) from the naming above, is that `docker pull jolibrain/deepdetect` without forcing the `:latest-gpu`, would pull the cpu version by default. I this is the case, I can see this as an issue, with users reporting they are not able to run on GPU (there'll be no warning, asking for gpu, would simply fall back on CPU).

**Comment by quadeare** [comment]
In this case, we can keep a split between CPU and GPU image :

Example (CPU) :

* jolibrain/deepdetect_cpu:latest
* jolibrain/deepdetect_cpu:latest-caffe-tf
* jolibrain/deepdetect_cpu:v0.9.4
* jolibrain/deepdetect_cpu:v0.9.4-caffe-tf

Example (GPU) :

* jolibrain/deepdetect_gpu:latest
* jolibrain/deepdetect_gpu:latest-caffe-tf
* jolibrain/deepdetect_gpu:v0.9.4
* jolibrain/deepdetect_gpu:v0.9.4-caffe-tf

I think it can be a good compromise.

**Comment by beniz** [comment]
OK, smart! 

Some builds have GPU 'flavors', and we should be able to remove that. This was due to some versions of CUDA not supporting some new cards. Basically, the trend is this: when new NVidia cards come out, there are supported only with the most recent version of CUDA. With docker + GPU, CUDA is actually within the docker image, so the base image can be changed to modify the cuda version.

**Comment by quadeare** [comment]
Well... we can support dynamic cuda version with one Dockerfile by using variable like this :

``` bash
docker build --build-arg DEEPDETECT_BUILD=caffe-tf \
--build-arg CUDA_VERSION=10.2-cudnn7 \
-t jolibrain/deepdetect_cpu --no-cache -f cpu.Dockerfile .
```

``` bash
# Default CUDA version
ARG  CUDA_VERSION=9.0-cudnn7
FROM nvidia/cuda:${CUDA_VERSION}-devel AS build
...
...
FROM nvidia/cuda:${CUDA_VERSION}-runtime
...
...
```

That's a bit tricky, but it works ;)

Official Docker documentation : https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

**Comment by quadeare** [comment]
I just fix cuda Docker image under Ubuntu 16.04 to avoid an unexpected upgrade from Nvidia team.
Commit : [348627d](https://github.com/jolibrain/deepdetect/pull/706/commits/348627d69b601cd68421359894bf5a06d7f1cb1b)

**Comment by beniz** [comment]
> ```shell
> docker build --build-arg DEEPDETECT_BUILD=caffe-tf \
> --build-arg CUDA_VERSION=10.2-cudnn7 \
> -t jolibrain/deepdetect_cpu --no-cache -f cpu.Dockerfile .
> ```

That'd be pretty cool!

**Comment by quadeare** [comment]
> > ```shell
> > docker build --build-arg DEEPDETECT_BUILD=caffe-tf \
> > --build-arg CUDA_VERSION=10.2-cudnn7 \
> > -t jolibrain/deepdetect_cpu --no-cache -f cpu.Dockerfile .
> > ```
> 
> That'd be pretty cool!

Feature implemented on commit [99c46d7](https://github.com/jolibrain/deepdetect/pull/706/commits/99c46d73de7921d150e5a8c782ef27b39408b062)

File: docker/build.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    esac
+
+}
+
+# Select build profile
+if [[ ${DEEPDETECT_ARCH} == "cpu" ]]; then
+    cpu_build
+elif [[ ${DEEPDETECT_ARCH} == "gpu" ]]; then
+    gpu_build
+fi
```
[the remark this answers is not in the document]
Comment (quadeare) [reply docker/build.sh]: Indeed, that's better ! Maybe we can put an interactive selector when DEEPDETECT_ARCH or DEEPDETECT_BUILD are empty.

Like this : https://linuxhint.com/bash_select_command/

File: docker/README.md [hunk 2]
```diff
@@ -112,10 +112,52 @@ docker run -d -p 8080:8080 -v /path/to/volume:/mnt jolibrain/deepdetect_cpu
...
+  * caffe-tf
+  * default
+* GPU
+  * caffe-cpu-tf
+  * caffe-tf
+  * caffe2
+  * p100
+  * volta
+  * default
+
```
[the remark this answers is not in the document]
Comment (quadeare) [reply docker/README.md]: Do you want to use this script standalone ? In this case, we probably need to put build.sh script at root.

File: docker/build.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (alx) [reply docker/build.sh]: Nice, can you add a `select` command when `DEEPDETECT_ARCH` is missing ? `DEEPDETECT_BUILD` can now use the default statement.

File: docker/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (alx) [reply docker/README.md]: Ah ok, error from my part, I didn't realize that this script was only used in Dockerfiles.

[the remark this answers is not in the document]
Comment (quadeare) [reply docker/README.md]: We can use this script on both (Dockerfile and standalone) so i'll move build script at root ;)

**Comment by quadeare** [comment]
I create a build folder on root project to make build simpler with and without Docker. Now, we can call this script without Docker in build folder. I adapt .gitignore to ignore all excepted README.md and build.sh script.

I you prefer create manually build folder, i can rollback this modification and put build.sh script at root.

But now, we've got a README.md in build folder probably a bit useless. Maybe we could merge it with main README.md (i don't know...).

Anyway, i add some documentation about build script.

File: docker/build.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (quadeare) [reply docker/build.sh]: Related to this feature : https://github.com/jolibrain/deepdetect/pull/706/commits/723e6012f2d988234b03fcfd4859c00ab0b4b9aa

**Comment by alx** [comment]
Hi @quadeare 

could you merge https://github.com/jolibrain/deepdetect/pull/710 in your fork ? It fixes a docker image build issue when multiple `gen_code` are given to `CUDA_ARCH` variable.

Thanks,

Alex

File: gpu.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# Default CUDA version
+ARG CUDA_VERSION=9.0-cudnn7
```
Comment (alx) [inline gpu.Dockerfile]: @fantes could you review the default cuda version defined here ?

It probably impacts `-DCUDA_ARCH` cmake options in `build/build.sh` script.

I've been able to compile on my setup using:

```
ARG CUDA_VERSION=10.1-cudnn7
```

File: build/build.sh [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    esac
+
+}
+
+gpu_build() {
+
+    case ${DEEPDETECT_BUILD} in
+
+    "caffe-cpu-tf")
+        cmake .. -DUSE_TF=ON -DUSE_TF_CPU_ONLY=ON -DUSE_CUDNN=ON -DUSE_XGBOOST=ON -DUSE_SIMSEARCH=ON -DUSE_TSNE=ON -DCUDA_ARCH="-gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_52,code=sm_52 -gencode arch=compute_53,code=sm_53 -gencode arch=compute_61,code=sm_61"
```
Comment (alx) [inline build/build.sh]: @fantes Could you review `-DCUDA_ARCH` options in multiple `DEEPDETECT_BUILD` targets ?

Thanks, Alex

**Comment by cchadowitz-pf** [comment]
As someone who has been building DeepDetect in a custom Dockerfile, it could be helpful to have docker images that do *not* include models (or maybe make it an option to do so) so that the images are even smaller. Then users can provide the models they want at runtime by mounting a models folder, or basing a custom image on this baseline DeepDetect no-models image.

**Comment by quadeare** [comment]
> Hi @quadeare
> 
> could you merge #710 in your fork ? It fixes a docker image build issue when multiple `gen_code` are given to `CUDA_ARCH` variable.
> 
> Thanks,
> 
> Alex

Done : e176cd6

**Comment by fantes** [comment]
the e176cd6 merges many stuff recently added to master, and not only 0f4bb0d, i am not sure that was what you intended to do. 

> > Hi @quadeare
> > could you merge #710 in your fork ? It fixes a docker image build issue when multiple `gen_code` are given to `CUDA_ARCH` variable.
> > Thanks,
> > Alex
> 
> Done : [e176cd6](https://github.com/jolibrain/deepdetect/commit/e176cd6acd793f3560967a8e5435705944a72143)

**Comment by quadeare** [comment]
> the [e176cd6](https://github.com/jolibrain/deepdetect/commit/e176cd6acd793f3560967a8e5435705944a72143) merges many stuff recently added to master, and not only [0f4bb0d](https://github.com/jolibrain/deepdetect/commit/0f4bb0d289339195abb925a8726176953a3cb4d2), i am not sure that was what you intended to do.
> 
> > > Hi @quadeare
> > > could you merge #710 in your fork ? It fixes a docker image build issue when multiple `gen_code` are given to `CUDA_ARCH` variable.
> > > Thanks,
> > > Alex
> > 
> > 
> > Done : [e176cd6](https://github.com/jolibrain/deepdetect/commit/e176cd6acd793f3560967a8e5435705944a72143)

ohhh... maybe you're right. You think i have to only cherrypick this commit ? : https://github.com/jolibrain/deepdetect/pull/710/commits/0f4bb0d289339195abb925a8726176953a3cb4d2

**Comment by alx** [comment]
> As someone who has been building DeepDetect in a custom Dockerfile, it could be helpful to have docker images that do _not_ include models (or maybe make it an option to do so) so that the images are even smaller. Then users can provide the models they want at runtime by mounting a models folder, or basing a custom image on this baseline DeepDetect no-models image.

@quadeare what's your opinion on this request ? would it be possible to have a `--build-arg` flag, or something similar, in order to enable/disable these model inclusions during image building ?

For history reference, these models have been included for users to have access to some simple examples without digging into Docker volume mounting.

**Comment by fantes** [comment]
> > the [e176cd6](https://github.com/jolibrain/deepdetect/commit/e176cd6acd793f3560967a8e5435705944a72143) merges many stuff recently added to master, and not only [0f4bb0d](https://github.com/jolibrain/deepdetect/commit/0f4bb0d289339195abb925a8726176953a3cb4d2), i am not sure that was what you intended to do.
> > > > Hi @quadeare
> > > > could you merge #710 in your fork ? It fixes a docker image build issue when multiple `gen_code` are given to `CUDA_ARCH` variable.
> > > > Thanks,
> > > > Alex
> > > 
> > > 
> > > Done : [e176cd6](https://github.com/jolibrain/deepdetect/commit/e176cd6acd793f3560967a8e5435705944a72143)
> 
> ohhh... maybe you're right. You think i have to only cherrypick this commit ? : [0f4bb0d](https://github.com/jolibrain/deepdetect/commit/0f4bb0d289339195abb925a8726176953a3cb4d2)

that seems better, or merge master before merging this branch, whatever you prefer :)

**Comment by quadeare** [comment]
> > As someone who has been building DeepDetect in a custom Dockerfile, it could be helpful to have docker images that do _not_ include models (or maybe make it an option to do so) so that the images are even smaller. Then users can provide the models they want at runtime by mounting a models folder, or basing a custom image on this baseline DeepDetect no-models image.
> 
> @quadeare what's your opinion on this request ? would it be possible to have a `--build-arg` flag, or something similar, in order to enable/disable these model inclusions during image building ?
> 
> For history reference, these models have been included for users to have access to some simple examples without digging into Docker volume mounting.

That's a great idea, but 

> > > the [e176cd6](https://github.com/jolibrain/deepdetect/commit/e176cd6acd793f3560967a8e5435705944a72143) merges many stuff recently added to master, and not only [0f4bb0d](https://github.com/jolibrain/deepdetect/commit/0f4bb0d289339195abb925a8726176953a3cb4d2), i am not sure that was what you intended to do.
> > > > > Hi @quadeare
> > > > > could you merge #710 in your fork ? It fixes a docker image build issue when multiple `gen_code` are given to `CUDA_ARCH` variable.
> > > > > Thanks,
> > > > > Alex
> > > > 
> > > > 
> > > > Done : [e176cd6](https://github.com/jolibrain/deepdetect/commit/e176cd6acd793f3560967a8e5435705944a72143)
> > 
> > 
> > ohhh... maybe you're right. You think i have to only cherrypick this commit ? : [0f4bb0d](https://github.com/jolibrain/deepdetect/commit/0f4bb0d289339195abb925a8726176953a3cb4d2)
> 
> that seems better, or merge master before merging this branch, whatever you prefer :)

Hum.... You think i can rewrite history to clean commits ? I'm on my repository, i don't think it will have any impact.

**Comment by fantes** [comment]
> > > As someone who has been building DeepDetect in a custom Dockerfile, it could be helpful to have docker images that do _not_ include models (or maybe make it an option to do so) so that the images are even smaller. Then users can provide the models they want at runtime by mounting a models folder, or basing a custom image on this baseline DeepDetect no-models image.
> > 
> > 
> > @quadeare what's your opinion on this request ? would it be possible to have a `--build-arg` flag, or something similar, in order to enable/disable these model inclusions during image building ?
> > For history reference, these models have been included for users to have access to some simple examples without digging into Docker volume mounting.
> 
> That's a great idea, but
> 
> > > > the [e176cd6](https://github.com/jolibrain/deepdetect/commit/e176cd6acd793f3560967a8e5435705944a72143) merges many stuff recently added to master, and not only [0f4bb0d](https://github.com/jolibrain/deepdetect/commit/0f4bb0d289339195abb925a8726176953a3cb4d2), i am not sure that was what you intended to do.
> > > > > > Hi @quadeare
> > > > > > could you merge #710 in your fork ? It fixes a docker image build issue when multiple `gen_code` are given to `CUDA_ARCH` variable.
> > > > > > Thanks,
> > > > > > Alex
> > > > > 
> > > > > 
> > > > > Done : [e176cd6](https://github.com/jolibrain/deepdetect/commit/e176cd6acd793f3560967a8e5435705944a72143)
> > > 
> > > 
> > > ohhh... maybe you're right. You think i have to only cherrypick this commit ? : [0f4bb0d](https://github.com/jolibrain/deepdetect/commit/0f4bb0d289339195abb925a8726176953a3cb4d2)
> > 
> > 
> > that seems better, or merge master before merging this branch, whatever you prefer :)
> 
> Hum.... You think i can rewrite history to clean commits ? I'm on my repository, i don't think it will have any impact.

usually, we do whatever we want on our own branches and then { rebase -i on master / squash everything / force push }  before merging to master, in order to keep a master history almost clean. Problems arise when there were too many merge / cherry-pick on one own branch, in such case "rebase -i" becomes very very messy . I am not good enough with git to tell for sure what is the best way to do :)

**Comment by quadeare** [comment]
> Hi @quadeare
> 
> could you merge #710 in your fork ? It fixes a docker image build issue when multiple `gen_code` are given to `CUDA_ARCH` variable.
> 
> Thanks,
> 
> Alex

@alx @fantes Ok, i'm sorry for my merging storm ! But that's good now, i only cherry-pick https://github.com/jolibrain/deepdetect/pull/710/commits/0f4bb0d289339195abb925a8726176953a3cb4d2 on my repository.

**Comment by quadeare** [comment]
> > As someone who has been building DeepDetect in a custom Dockerfile, it could be helpful to have docker images that do _not_ include models (or maybe make it an option to do so) so that the images are even smaller. Then users can provide the models they want at runtime by mounting a models folder, or basing a custom image on this baseline DeepDetect no-models image.
> 
> @quadeare what's your opinion on this request ? would it be possible to have a `--build-arg` flag, or something similar, in order to enable/disable these model inclusions during image building ?
> 
> For history reference, these models have been included for users to have access to some simple examples without digging into Docker volume mounting.

That's a great idea, but we're not going to save a lot of disk space.

In fact, on CPU/GPU images, models take only about 160MB.
But, technically, we can pass argument to enable/disable models downloading during build.

If you choose to not include model downloading in Docker build, we have to decide which Docker tag we have to use on both case (with and without models).

**Comment by beniz** [comment]
> If you choose to not include model downloading in Docker build, we have to decide which Docker tag we have to use on both case (with and without models).

Good point. I guess we can include the models as default, and have an option in the script for users like @cchadowitz-pf who do know how to build a custom image. Any thoughts/preferences on this @cchadowitz-pf ?

**Comment by cchadowitz-pf** [comment]
I think that'd be fine. I know the models aren't terribly large, and it is definitely useful for users to have out-of-the-box models to try out examples, but it would definitely be great to have an official base image to work from. 

Another alternative is to include a script in the image that pulls the models that are currently included in the image when the image is run, at which point images that are based on this can simply override the default startup command. I believe DeepDetect also supports loading models from URLs as well, right? So it could even just simply start with the models to be initialized via URL. 

I suppose both these options do not persist the models though, like they would be if they're included natively in the image, or if a volume mount is used.

**Comment by quadeare** [comment]
I can't choose between both solutions. Like i said, models are not heavy so, i think we can keep it inside Docker images.

Can i do anything else on this issue ? Have you got another suggestion ?

**Comment by beniz** [comment]
Hi @quadeare I'd say the suggestion by @cchadowitz-pf is interesting: we could include a script that defaults to pulling the existing demo models. This script could be overriden to build custom images that would include custom models. This is useful to companies and organizations that build their own images.

@quadeare what would be your recommendation, adding this to the `build.sh` script ? Thanks!

**Comment by quadeare** [comment]
Due to the multi-stage build, i have to create another script to download models in Docker image (build/get_models.sh).

By default, models are downloaded in Docker image, but we can pass argument to disable this feature (DEEPDETECT_DEFAULT_MODELS=false). I had some documentation in README.md about this.

As expected, we can't save a lot of space (about 160MB) :

* CPU image : 
```bash
flacrampe_ol@neptune2:~/deepdetect_docker/build$ docker images | grep cpu
optimized/deepdetect_cpu          without-models                   d6bec7c071d9        23 hours ago        763MB
optimized/deepdetect_cpu          latest                           8c81f2d63e6f        24 hours ago        920MB
```

* GPU image : 

```bash
flacrampe_ol@neptune2:~/deepdetect_docker/build$ docker images | grep gpu
optimized/deepdetect_gpu          latest                           00678b6041b9        21 hours ago         2.44GB
optimized/deepdetect_gpu          without-models                   e67150e53795        22 hours ago         2.29GB

```

**Comment by cchadowitz-pf** [comment]
Very much appreciate this work and effort! I know it's not a lot of space, but every bit (hah...) counts in my case :)

**Comment by beniz** [comment]
@quadeare brilliant!

**Comment by fantes** [comment]
here is the official doc for cuda_arch flags : https://docs.nvidia.com/cuda/turing-compatibility-guide/index.html

File: build/build.sh [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        ;;
+    esac
+
+}
+
+gpu_build() {
+
+    case ${DEEPDETECT_BUILD} in
+
+    "caffe-cpu-tf")
```
Comment (alx) [inline build/build.sh]: @quadeare this build case must be moved to `cpu_build()`

File: build/build.sh (hunk 4 above)
Comment (alx) [reply build/build.sh]: https://docs.nvidia.com/cuda/turing-compatibility-guide/index.html

File: gpu.Dockerfile (hunk 3 above)
Comment (alx) [reply gpu.Dockerfile]: > à partir de 10.0, il y support pour sm_75, qui concerne les dernières cartes nvidia , cad RTX 2080, Titan RTX, Quadro R8000
> forcer 10.0 / 10.1 par rapport à 9.0 n'enlève pas de cartes par contre

File: build/build.sh (hunk 4 above)
Comment (alx) [reply build/build.sh]: > 30 et 35 concernent des vieilles vieilles cartes plus supportées je pense , et 53 et 62 n'ont jamais existé je pense

Flags to use depending on chosen `CUDA_VERSION`:

* Cuda 9

```
  -gencode=arch=compute_50,code=sm_50
  -gencode=arch=compute_52,code=sm_52
  -gencode=arch=compute_60,code=sm_60
  -gencode=arch=compute_61,code=sm_61
  -gencode=arch=compute_70,code=sm_70
  -gencode=arch=compute_70,code=compute_70
```

* Cuda 10

```
  -gencode=arch=compute_50,code=sm_50
  -gencode=arch=compute_52,code=sm_52
  -gencode=arch=compute_60,code=sm_60
  -gencode=arch=compute_61,code=sm_61
  -gencode=arch=compute_70,code=sm_70
  -gencode=arch=compute_75,code=sm_75
  -gencode=arch=compute_75,code=compute_75
```

File: build/build.sh (hunk 5 above)
Comment (quadeare) [reply build/build.sh]: Arf... My bad, that's fixed : https://github.com/jolibrain/deepdetect/pull/706/commits/08e8a1fbd1f3b0c525b3e203b861487f750e25e3

**Comment by alx** [comment]
Hi @quadeare 

I've tried to build all docker images using various config possibilities, and still got some errors on some combination:

- `cpu.Dockerfile` with `DEEPDETECT_BUILD=armv7`, does it need to build it on an armv7 hardware? [deepdetect_cpu_armv7_optimized.log](https://github.com/jolibrain/deepdetect/files/4375584/deepdetect_cpu_armv7_optimized.log)
- `cpu.Dockerfile` with `DEEPDETECT_BUILD=caffe-tf` [deepdetect_cpu_caffe-tf_optimized.log](https://github.com/jolibrain/deepdetect/files/4375582/deepdetect_cpu_caffe-tf_optimized.log)
- `gpu.Dockerfile` with `DEEPDETECT_BUILD=caffe-tf` [deepdetect_gpu_optimized_caffe-tf.log](https://github.com/jolibrain/deepdetect/files/4375580/deepdetect_gpu_optimized_caffe-tf.log)
- `gpu.Dockerfile` with `DEEPDETECT_BUILD=caffe2` [deepdetect_gpu_optimized_caffe2.log](https://github.com/jolibrain/deepdetect/files/4375581/deepdetect_gpu_optimized_caffe2.log)

Script to build all docker images:
[build_all_docker_images.sh.gz](https://github.com/jolibrain/deepdetect/files/4375611/build_all_docker_images.sh.gz)

Could you have a look ?

Thanks,

Alex

**Comment by quadeare** [comment]
First, thank you for your review and logs files.

> cpu.Dockerfile with DEEPDETECT_BUILD=armv7, does it need to build it on an armv7 hardware? deepdetect_cpu_armv7_optimized.log

Build script is probably a bit confused. That's a standalone build script, so we have all CPU profiles.
If you try to build `cpu.Dockerfile` with `DEEPDETECT_BUILD=armv7` arg, you'll build an `armv7` version under `x86_64` arch. In this case, it crashes.

> cpu.Dockerfile with DEEPDETECT_BUILD=caffe-tf deepdetect_cpu_caffe-tf_optimized.log
> gpu.Dockerfile with DEEPDETECT_BUILD=caffe-tf deepdetect_gpu_optimized_caffe-tf.log

Arf... bazel dependency was missing, i forget to add it in Dockerfile.
I'll test build and run with this build-arg.

> gpu.Dockerfile with DEEPDETECT_BUILD=caffe2 deepdetect_gpu_optimized_caffe2.log

Bazel is also required but there is another error during a git clone (maybe caused by network hangup, i'll check it)

**Comment by quadeare** [comment]
Unfortunately, i've got same errors with deepdetect master branch.

* Build Dockerfile in `docker/cpu-caffe-tf` : [build-cpu-caffe-tf.log.log](https://github.com/jolibrain/deepdetect/files/4379687/build-cpu-caffe-tf.log.log)

* Build Dockerfile in `docker/gpu-caffe-tf` : [build-gpu-caffe-tf.log](https://github.com/jolibrain/deepdetect/files/4379689/build-gpu-caffe-tf.log)

* Build Dockerfile in `docker/gpu-caffe2` : [build-gpu-caffe2.log](https://github.com/jolibrain/deepdetect/files/4379787/build-gpu-caffe2.log)

If you have any clues, ping me ;)

**Pull request merged**